app/services/fact_cleaner.rb

```diff
@@ -32,25 +32,27 @@ def delete_facts_names_values(fact_name_ids)
 
   def delete_orphaned_facts
     logger.debug "Cleaning leaf orphaned facts"
-    leaf_ids = []
-    FactName.leaves.includes(:fact_values).reorder(nil).find_in_batches(:batch_size => @batch_size) do |batch|
+    FactName.leaves.preload(:fact_values).reorder(nil).find_in_batches(:batch_size => @batch_size) do |batch|
       fact_name_ids = batch.select { |fact| fact.fact_values.empty? }
-      @deleted_count += delete_facts_names_values(fact_name_ids)
-      leaf_ids += fact_name_ids
+      @deleted_count += log_removed(FactName.unscoped.where(:id => fact_name_ids).delete_all, "fact name")
     end
 
     logger.debug "Cleaning compose orphaned facts"
-    FactName.composes.reorder(nil).find_in_batches(:batch_size => @batch_size) do |batch|
-      fact_name_ids = batch.select { |compose| compose.descendants.all? { |fact| leaf_ids.include?(fact.id) } }.map(&:id)
-      @deleted_count += delete_facts_names_values(fact_name_ids)
+    live_ancestors = Set.new
+    FactName.leaves.reorder(nil).select(FactName.ancestry_base_class.ancestry_column, :id).find_each(:batch_size => @batch_size) do |leaf|
+      live_ancestors.merge(leaf.ancestor_ids)
+    end
+    FactName.composes.reorder(nil).where.not(id: live_ancestors).select(:id).find_in_batches(:batch_size => @batch_size) do |batch|
```

Comment:
We can improve it even further by using `.in_batches.pluck` here. We don't need the model object at all

Reply:
I had no idea this exists.

https://github.com/rails/rails/blob/v5.0.0/activerecord/lib/active_record/relation/batches.rb#L184

This is a very good memory optimization which is worth looking into - array of integers will be much more effective than AR instances. Could you incorporate this change?

Reply:
i tired that now and looks like it actually doubles the queries- it does one query for the ids to get the batch, then another to pluck the id from all the records in the batch

Reply:
specifically this is the code: https://github.com/rails/rails/blob/v5.0.0/activerecord/lib/active_record/relation/batches.rb#L205-L206 
doesn't look like there is an easy way of getting just the ids without doing our own batch handling.